be/src/runtime/runtime_filter_cache.cpp

```diff
@@ -0,0 +1,217 @@
+    for (auto i = 0; i < _num_slots; ++i) {
+        auto& mutex = _event_mutexes[i];
+        auto& map = _event_maps[i];
+        std::unique_lock write_lock(mutex);
+        for (auto it = map.begin(); it != map.end();) {
+            if (it->second->is_expired()) {
+                it = map.erase(it);
+            } else {
+                ++it;
+            }
+        }
+    }
+}
+void RuntimeFilterCache::_clean_filters() {
+    for (auto i = 0; i < _num_slots; ++i) {
+        auto& mutex = _mutexes[i];
+        auto& map = _filter_maps[i];
+        std::unique_lock write_lock(mutex);
+        for (auto it = map.begin(); it != map.end();) {
+            if (it->second->is_expired()) {
+                it = map.erase(it);
+            } else {
+                ++it;
+            }
+        }
+    }
+}
+void RuntimeFilterCache::put_if_absent(const TUniqueId& query_id, int filter_id, const JoinRuntimeFilterPtr& filter) {
+    const auto slot_idx = std::hash<size_t>()(query_id.lo) & _slot_mask;
+    auto& mutex = _mutexes[slot_idx];
+    auto& map = _filter_maps[slot_idx];
+    _cache_times.fetch_add(1);
+    std::unique_lock write_lock(mutex);
+    if (!map.count(query_id)) {
+        map[query_id] = std::make_shared<RfCacheValue>();
+    }
+    auto& cached = map[query_id];
+    cached->put_if_absent(filter_id, filter);
+}
+
+JoinRuntimeFilterPtr RuntimeFilterCache::get(const TUniqueId& query_id, int filter_id) {
+    const auto slot_idx = std::hash<size_t>()(query_id.lo) & _slot_mask;
+    auto& mutex = _mutexes[slot_idx];
+    auto& map = _filter_maps[slot_idx];
+    std::shared_lock read_lock(mutex);
+    auto it = map.find(query_id);
+    if (it == map.end()) {
+        return nullptr;
+    } else {
+        auto filter = it->second->get(filter_id);
+        _use_times.fetch_add(filter != nullptr);
+        return filter;
+    }
+}
+
+void RuntimeFilterCache::add_rf_event(const TUniqueId& query_id, int filter_id, std::string&& msg) {
+    const auto slot_idx = std::hash<size_t>()(query_id.lo) & _slot_mask;
+    auto& mutex = _event_mutexes[slot_idx];
+    auto& map = _event_maps[slot_idx];
+    std::unique_lock write_lock(mutex);
+    if (!map.count(query_id)) {
+        map[query_id] = std::make_shared<RfEventValue>();
+    }
+    map[query_id]->add_rf_event(query_id, filter_id, std::move(msg));
+}
+
+std::unordered_map<std::string, std::list<std::string>> RuntimeFilterCache::get_events() {
+    std::unordered_map<std::string, std::list<std::string>> events;
+    for (auto i = 0; i < _num_slots; ++i) {
+        auto& mutex = _event_mutexes[i];
+        auto& map = _event_maps[i];
+        std::shared_lock read_lock(mutex);
+        for (auto it = map.begin(); it != map.end(); ++it) {
+            events[print_id(it->first)] = it->second->get_events();
+        }
+    }
+    return events;
+}
+
+void RuntimeFilterCache::remove(const TUniqueId& query_id) {
+    const auto slot_idx = std::hash<size_t>()(query_id.lo) & _slot_mask;
```

Comment:
Every method calculates `slot_idx`, how about abstracting it to a method like `size_t _slot_idx(const TUniqueId& query_id)`?

Reply:
ok